### PR TITLE
Heapster Array Prover Redux

### PR DESCRIPTION
**Pull request by abakst**
+ Reimplementation of Array prover logic in Heapster
+ Adds test framework for Heapster implication prover plus Array testcases

### CODE REVIEW

File: heapster-saw/src/Verifier/SAW/Heapster/Permissions.hs [hunk 1]
```diff
@@ -4648,6 +4675,30 @@ cellOffsetLLVMArrayBorrow off (FieldBorrow ix) =
 cellOffsetLLVMArrayBorrow off (RangeBorrow rng) =
   RangeBorrow $ offsetBVRange off rng
 
+llvmArrayBorrowCells :: (KnownNat w, 1 <= w) => LLVMArrayBorrow w -> BVRange w
```
Comment (eddywestbrook) [inline heapster-saw/src/Verifier/SAW/Heapster/Permissions.hs:4678]: Haddock

File: heapster-saw/src/Verifier/SAW/Heapster/Permissions.hs [hunk 2]
```diff
@@ -4855,6 +4936,20 @@ llvmAtomicPermOverlapsRange rng (Perm_LLVMBlock bp) =
   bvRangesOverlap rng (llvmBlockRange bp)
 llvmAtomicPermOverlapsRange _ _ = False
 
+-- | Test if an atomic LLVM permission has a range that overlaps with (in the
```
Comment (eddywestbrook) [inline heapster-saw/src/Verifier/SAW/Heapster/Permissions.hs:4939]: Small nitpick, but I would prefer to say "could overlap" in the comments

File: heapster-saw/src/Verifier/SAW/Heapster/Implication.hs [hunk 3]
```diff
@@ -6318,11 +6356,153 @@ proveVarLLVMFieldH2 x p _ mb_fp =
...
+          | bvCouldBeInRange (llvmBlockOffset bp) range -> Just bp
+        _ -> Nothing
+
+-- | Given an array permission ap and permission p, calculate the borrow of ap
+-- corresponding to p, plus the range of bytes covered by p.
+permToBorrowRange ::
+  forall w. (1 <= w, KnownNat w) =>
+  LLVMArrayPerm w ->
+  AtomicPerm (LLVMPointerType w) ->
+  Maybe (LLVMArrayBorrow w, BVRange w)
```
Comment (abakst) [inline heapster-saw/src/Verifier/SAW/Heapster/Implication.hs:6397]: Move this to Permissions.hs and only compute the Borrow

File: heapster-saw/src/Verifier/SAW/Heapster/Implication.hs [hunk 4]
```diff
@@ -6318,11 +6356,153 @@ proveVarLLVMFieldH2 x p _ mb_fp =
...
+
+-- | Given a list ps of permissions, find the subseqeuences of ps
+-- that could cover the given array permission. Also returns the permissions
+-- corresponding to the  given ranges.
+gatherRangesForArray ::
+  forall w.
+  (1 <= w, KnownNat w) =>
+  [AtomicPerm (LLVMPointerType w)] ->
+  LLVMArrayPerm w ->
+  [[(Maybe (AtomicPerm (LLVMPointerType w)), (LLVMArrayBorrow w, BVRange w))]]
```
Comment (abakst) [inline heapster-saw/src/Verifier/SAW/Heapster/Implication.hs:6423]: Remove the BVRange from the returned value

File: heapster-saw/src/Verifier/SAW/Heapster/Implication.hs [hunk 5]
```diff
@@ -6318,11 +6356,153 @@ proveVarLLVMFieldH2 x p _ mb_fp =
...
+-- the stack.
+proveSomeBlockM :: forall w r s vars ps.
+  (1 <= w, KnownNat w, NuMatchingAny1 r) =>
+  ExprVar (LLVMPointerType w) ->
+  [AtomicPerm (LLVMPointerType w)] ->
+  [AtomicPerm (LLVMPointerType w)] ->
+  BVRange w ->
+  ImplM vars s r (ps :> LLVMPointerType w) (ps :> LLVMPointerType w) (LLVMBlockPerm w)
+proveSomeBlockM x ps ps' range
+  | bp:_ <- catMaybes (couldProve <$> ps') =
```
[the remark this answers is not in the document]
Comment (eddywestbrook) [reply heapster-saw/src/Verifier/SAW/Heapster/Implication.hs:6375]: Never mind, this doesn't make sense...

File: heapster-saw/src/Verifier/SAW/Heapster/Permissions.hs [hunk 6]
```diff
@@ -3715,6 +3723,19 @@ llvmPtr0EqExPerm :: (1 <= w, KnownNat w, 1 <= sz, KnownNat sz) => prx sz ->
                     (ValuePerm (LLVMPointerType w))
 llvmPtr0EqExPerm = fmap (ValPerm_Conj1 . Perm_LLVMField) . llvmPtr0EqEx
 
+llvmExBlock :: (1 <= w, KnownNat w) =>
```
[the remark this answers is not in the document]
Comment (abakst) [reply heapster-saw/src/Verifier/SAW/Heapster/Permissions.hs:3726]: Oops, this one should have been garbage collected!

File: heapster-saw/src/Verifier/SAW/Heapster/Implication.hs [hunk 7]
```diff
@@ -1912,7 +1925,11 @@ simplImplIn (SImpl_LLVMArrayReturn x ap ret_ap) =
     distPerms2 x (ValPerm_Conj [Perm_LLVMArray ret_ap])
     x (ValPerm_Conj [Perm_LLVMArray ap])
   else
-    error "simplImplIn: SImpl_LLVMArrayReturn: array not being borrowed or not a sub-array"
+    error ("simplImplIn: SImpl_LLVMArrayReturn: array not being borrowed or not a sub-array:\n" ++
+            renderDoc (
```
[the remark this answers is not in the document]
Comment (abakst) [reply heapster-saw/src/Verifier/SAW/Heapster/Implication.hs:1934]: Actually I had only added this for my own debugging and missed it when I initially reviewed the diffs. @eddywestbrook do you think it's useful to leave in, given that it has to use `emptyPPInfo`?

[the remark this answers is not in the document]
Comment (eddywestbrook) [reply heapster-saw/src/Verifier/SAW/Heapster/Implication.hs:1934]: You're right, you should just take it out

**Review by eddywestbrook — APPROVED** [review]
Looks good, thanks Alex!

**Pull request merged**